urlac47192b66_ggrandes_bouncer_tgz-pJ8-org_javastack_bouncer_ClusterTestJ8/src/org/javastack/bouncer/MuxServer.java

```diff
@@ -0,0 +1,566 @@
+		}
+
+		@Override
+		public void run() {
+			Log.info(this.getClass().getSimpleName() + "::run socket: " + sock);
+			long muxNopKeepAlive = System.currentTimeMillis();
+			while (!shutdown || !mapRemotes.isEmpty()) {
+				try {
+					final MuxPacket msg = context.allocateMuxPacket();
+					if (seal != null) {
+						// AES encryption
+						final byte[] encoded = IOHelper.fromWireWithHeader(is);
+						final byte[] decoded = seal.decode(encoded, 0, encoded.length);
+						final ByteArrayInputStream bais = new ByteArrayInputStream(decoded);
+						msg.fromWire(bais);
+					} else {
+						msg.fromWire(is);
+					}
+					context.getStatistics().incInMsgs().incInBytes(msg.getBufferLen());
+					router.onReceiveFromLocal(this, msg);
+					context.releaseMuxPacket(msg);
+				} catch (SocketTimeoutException e) {
+					final long now = System.currentTimeMillis();
+					if ((muxNopKeepAlive + Constants.MUX_KEEP_ALIVE) < now) {
+						Log.debug(this.getClass().getSimpleName() + " " + e.toString());
+						sendNOP();
+						muxNopKeepAlive = now;
+					}
+					continue;
+				} catch (EOFException e) {
+					break;
+				} catch (IOException e) {
+					if (!sock.isClosed() && !shutdown) {
+						Log.error(this.getClass().getSimpleName() + " " + e.toString());
+					}
+					break;
+				} catch (GeneralSecurityException e) {
+					Log.error(this.getClass().getSimpleName() + " " + e.toString());
+					doSleep(1000);
+					break;
+				} catch (Exception e) {
+					Log.error(this.getClass().getSimpleName() + " Generic exception", e);
+					break;
+				}
+			}
+			// Close all
+			close();
+			closeRemote();
+			synchronized (mapRemotes) { // Remotes are RAW
+				for (MuxServerRemote r : mapRemotes.values()) {
+					r.setShutdown();
+				}
+				mapRemotes.clear();
+			}
+			Log.info(this.getClass().getSimpleName() + " await end");
+			context.awaitShutdown(this);
+			Log.info(this.getClass().getSimpleName() + " end");
+			local = null;
+		}
+	}
+
+	class MuxServerRemote extends MuxServerConnection { // Remote is RAW
+		final Semaphore isLocked = new Semaphore(0); // Begin Locked
+		final ArrayBlockingQueue<RawPacket> queue = new ArrayBlockingQueue<RawPacket>(
+				Constants.IO_BUFFERS << 1);
+		final int id;
+		long keepalive = System.currentTimeMillis();
+
+		MuxServerRemote(final Socket sock, final InboundAddress inboundAddress) throws IOException {
+			super(sock, inboundAddress);
+			id = sock.getPort();
+		}
+
+		void unlock(final int size) {
+			isLocked.release(size);
+		}
+
+		boolean lock(final int size) throws InterruptedException {
+			return isLocked.tryAcquire(size, 3000, TimeUnit.MILLISECONDS);
+		}
+
+		int getId() {
+			return id;
+		}
+
+		void sendQueueRemote(final RawPacket msg) throws IOException {
+			try {
+				while (!queue.offer(msg, 1000, TimeUnit.MILLISECONDS)) {
+					if (shutdown)
+						break;
+				}
+				keepalive = System.currentTimeMillis();
+			} catch (InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)